@xen-orchestra/backups/RemoteAdapter.js

```diff
@@ -774,7 +774,46 @@ class RemoteAdapter {
     // _filename is a private field used to compute the backup id
     //
     // it's enumerable to make it cacheable
-    return { ...JSON.parse(await this._handler.readFile(path)), _filename: path }
+    const metadata = { ...JSON.parse(await this._handler.readFile(path)), _filename: path }
+
+    // older xenserver version store boolean as int, and it breaks on restore
+    if (typeof metadata.vm.is_a_template === 'number') {
+      const properties = {
+        vbds: ['bootable', 'unpluggable', 'storage_lock', 'empty', 'currently_attached'],
+        vdis: [
+          'sharable',
+          'read_only',
+          'storage_lock',
+          'managed',
+          'missing',
+          'is_a_snapshot',
+          'allow_caching',
+          'metadata_latest',
+        ],
+        vifs: ['currently_attached', 'MAC_autogenerated'],
+        vm: ['is_a_template', 'is_control_domain', 'ha_always_run', 'is_a_snapshot', 'is_snapshot_from_vmpp'],
+        vmSnapshot: ['is_a_template', 'is_control_domain', 'ha_always_run', 'is_snapshot_from_vmpp'],
+      }
+
+      function clean(obj, properties) {
+        properties.forEach(key => {
+          if (typeof obj[key] === 'number') {
+            obj[key] = obj[key] === 1
+          }
+        })
+      }
+
+      for (const [key, propertiesInKey] of Object.entries(properties)) {
+        if (metadata[key] !== undefined) {
+          const value = metadata[key]
+          const isCollection = Object.keys(value).some(subKey => subKey.startsWith('OpaqueRef:'))
+          isCollection
+            ? Object.keys(metadata[key]).forEach(subKey => clean(metadata[key][subKey], propertiesInKey))
+            : clean(metadata[key], propertiesInKey)
+        }
```

Comment:
Various suggestions, take the ones you like :slightly_smiling_face: 

- more explicit function name
- less key vs prop
- simpler collection test
- avoid using ternary operator when value is not needed
- always use `value` instead of `metadata[key]`

```suggestion
      function fixBooleans(obj, booleanProps) {
        booleanProps.forEach(prop => {
          if (typeof obj[prop] === 'number') {
            obj[prop] = obj[prop] === 1
          }
        })
      }

      for (const [key, booleanProps] of Object.entries(properties)) {
        if (metadata[key] !== undefined) {
          const value = metadata[key]
          const isCollection = key[key.length - 1] === 's'
          if (isCollection) {
            Object.values(value).forEach(item => fixBooleans(item, booleanProps))
          } else {
            fixBooleans(value, booleanProps)
          }
        }
```

Reply:
I applied the suggestions except from the collection test . I think testing the label is not better than testing opaque ref